Updated string interpolation tutorial

pkulikov opened the pull request:
Updated the [string interpolation tutorial](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/string-interpolation):
- use Q&A format
- use interactive code snippets and skip creation of the separate console app

This PR is dependent on dotnet/samples#47

Related to #3285

pkulikov committed:
2 out of 6 complete commit

pkulikov committed:
Sections 3 and 4 added

pkulikov committed:
Added culture-related sections

pkulikov committed:
Finalized tutorial

pkulikov committed:
Merged in master

pkulikov committed:
Fixed build warning

rpetrusha commented:
Closing and reopening to kick off new build.

pkulikov committed:
Link tutorial from the quickstart

pkulikov committed:
Merge local to remote

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+author: pkulikov
+ms.date: 05/07/2018
 ---
+# String interpolation in C# #
```
Since you've made the C# examples interactive, the csharp-interactive-note.md file should be included.

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
 
+## How to control the field width and alignment of the formatted interpolated expression
 
+You specify the minimum field width in which string representation of the expression result is inserted and the text alignment within that field by following the interpolated expression with a comma (",") and the constant expression that defines alignment:
```
- Somewhat more concise: "You specify the minimum field width and the alignment of the string produced by the interpolated expression by following...
- the constant expression --> an expression (constant here is capable of misinterpretation)
"that defines alignment" is unnecessary, since you've already said it defines alignment and field width.

pkulikov replied:
>the constant expression --> an expression (constant here is capable of misinterpretation)

How can it be misinterpreted? I then try to clarify that. Just *an expression* is not correct: an expression must be possible to evaluate at the compile time. Thus I've used *constant* term.

rpetrusha replied:
The confusion about constant is that it's typically interpreted as a fixed (not variable) value. Your example, though, uses variables to define field width.

pkulikov replied:
In my example, they are not variables, they are local constants. I cannot change their values after declaration.

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
````diff
@@ -1,122 +1,98 @@
+```
 
-With <xref:System.String.Format%2A?displayProperty=nameWithType>, you specify "placeholders" in a string that are replaced by the arguments following the string. For instance:
+If the alignment value is positive, the formatted string representation of the expression result is right-aligned; if negative, it's left-aligned.
````
alignment can be italicized 
perhaps instead of "the formatted string representation of the expression result", just "the expression result"

pkulikov replied:
I've phrased 
> ...the formatted string representation of the expression result...

to stress the fact that the expression result is first formatted to text and only then alignment is applied.

rpetrusha replied:
OK, that's fine.

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+The following example shows how to specify alignment:
 
-[!code-csharp[Interpolation example](../../../samples/snippets/csharp/new-in-6/string-interpolation.cs#InterpolationExample)]  
+[!code-csharp-interactive[alignment example](~/samples/snippets/csharp/tutorials/string-interpolation/Program.cs#3)]
```
- I'm not sure what the "|" characters accomplish. (Were you thinking of a markdown table?) I think that I'd remove them
- NameAlignment should be 10 or greater, since "Arithmetic" has 10 characters.

pkulikov replied:
I find pipes ("|") in output make it easier to visualize alignment. For example, see the [current version of the tutorial](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/string-interpolation#how-string-interpolation-works), the example with two numeric columns: from the code, the right column is aligned to the right, while from output, it looks as if it's left-aligned. Just not a good choice of the field width makes the example confusing. With the pipes in output example would be clearer.

As for the `NameAlignment` value, it's an intentional choice in order to show that the *minimum* field width is specified and what happens if the formatted text is longer.

rpetrusha replied:
In that case, it would be good to explain the pipes. Somehow with pipes in text, I (at least) expect proper alignment.

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
````diff
@@ -1,122 +1,98 @@
-This is line number 4
-This is line number 5
-```
+[Verbatim](../language-reference/tokens/verbatim.md) interpolated strings start with the `$` character followed by the `@` character.
````
This needs to have a short explanation of what a verbatim string literal is.

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+## How to use a ternary conditional operator in an interpolated expression
 
-[!code-csharp[Interpolation formatting example](../../../samples/snippets/csharp/new-in-6/string-interpolation.cs#InterpolationFormattingExample)]  
+As the colon (":") has special meaning in an item with an interpolated expression, in order to use a [conditional operator](../language-reference/operators/conditional-operator.md) in an expression, enclose it in parentheses, as the following example shows:
```
This should start with a one-sentence explanation of the ? : language element.

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+## How to create a culture-specific result string with string interpolation
 
-If a variable name is not found, then a compile-time error is generated.
+By default, an interpolated string resolves to the result string with the use of the current culture defined by the <xref:System.Globalization.CultureInfo.CurrentCulture> property. Use implicit conversion of an interpolated string to a <xref:System.FormattableString?displayProperty=nameWithType> instance and its <xref:System.FormattableString.ToString(System.IFormatProvider)> method to create a c ...
```
- The first sentence is a bit awkward. Perhaps, "an interpolated string uses the current culture defined by the \<xref:System.Globalization.CultureInfo.CurrentCulture?displayProperty=nameWithType> property for all formatting operations." 
- "and its" --> and call its

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
````diff
@@ -1,122 +1,98 @@
-var adj = "quick";
-Console.WriteLine(localizeMe);
-```
+You can use one <xref:System.FormattableString> instance to generate multiple result strings for various cultures.
````
perhaps preface this with, "As the example shows,"

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+## How to create the result string for the invariant culture
 
-## Localization and Internationalization
+Along with the <xref:System.FormattableString.ToString(System.IFormatProvider)?displayProperty=nameWithType> method, you can use the static <xref:System.FormattableString.Invariant%2A?displayProperty=nameWithType> method to resolve an interpolated string into the result string for the <xref:System.Globalization.CultureInfo.InvariantCulture>. The following example shows how to do that:
```
the result string --> a result string

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+## Conclusion
 
-[!code-csharp[Interpolation internationalization example](../../../samples/snippets/csharp/new-in-6/string-interpolation.cs#InterpolationInternationalizationExample)]  
+This tutorial describes the most common scenarios of the string interpolation usage. For more information about string interpolation, see the [String interpolation](../language-reference/tokens/interpolated.md) topic. For more information about formatting types in .NET, see the [Formatting Types in .NET](../../standard/base-types/formatting-types.md) and [Composite formatting](../../standard/base ...
```
the string interpolation --> string interpolation

rpetrusha reviewed (changes requested):
Thanks for submitting this PR, @pkulikov. Your PR significantly improves the tutorial. I've left a number of comments, including one related to the code example.

pkulikov committed:
Addressed nits

rpetrusha commented:
@pkulikov, thanks for the detailed explanation. 

At the risk of making the string interpolation tutorial less introductory than the other tutorials that it's grouped with, what I think that you should, rather than add the prerequisites section and address a number of my other comments, is clearly specify the audience and the non-audience for the tutorial at the beginning. Indicate that this is a tutorial for intermediate to advanced .NET developers who either are unfamiliar with string interpolation or with FormattableString. For new .NET developers, or those who are very inexperienced with formatting, I'd probably add a note box with links that direct them to the quickstart or the local quickstart. 

Does that help?

rpetrusha commented on docs/csharp/tutorials/string-interpolation.md:
```diff
@@ -1,122 +1,98 @@
+The [string interpolation](../language-reference/tokens/interpolated.md) feature is built on top of the [composite formatting](../../standard/base-types/composite-formatting.md) feature and provides a more readable and convenient syntax to format string output.
 
-String Interpolation is the way that placeholders in a string are replaced by the value of a string variable. Before C# 6, the way to do this is with <xref:System.String.Format%2A?displayProperty=nameWithType>. This works okay, but since it uses numbered placeholders, it can be harder to read and more verbose.
+To identify a string literal as an interpolated string, prepend it with the `$` symbol. You can embed any valid C# expression that returns a value into an interpolated string:
```
I wasn't thinking of FormattableString here, and I think that that's a complexity that is best introduced later. Here, the issue is simple interpolated string functionality.

rpetrusha reviewed (commented):
You'd responded specifically to a few of my comments, @pkulikov, so I thought I'd respond to them. You can ignore my introductory tutorial-style remarks, except that I would still explain what a verbatim string literal is. (My general sense is that many developers who should know what it is don't.)

pkulikov committed:
Addressed feedback on tutorial

pkulikov commented:
@rpetrusha I've addressed remaining feedback, please review the updated version. Thank you for letting me experiment with the tutorial format, I appreciate that.

rpetrusha reviewed (approved):
Thanks, @pkulikov. You've done a really outstanding job in revising this documentation. I'll merge your PR now.
